c2hs crashes on invalid input

I had written this c2hs code;

```haskell
-- | Validate the key of a metadata pair.
{#fun pure grpc_header_key_is_legal as isKeyValid
  { 'useAsCStringLen'* `B.ByteString'&} -> `Bool' #}
```

But the C code had been updated from `char*, size_t` to `grpc_slice`;
```c
GRPCAPI int grpc_header_key_is_legal(grpc_slice slice);
```

Thus c2hs expects two arguments in the C function, but there was only one.
Naturally c2hs needs to fail generating the bindings, but the failure should be informative.

Currently it crashes at runtime with an error message which doesn't tell you where in your code you have a problem;

```
c2hs: src/C2HS/Gen/Bind.hs:(1426,5)-(1444,66): Non-exhaustive patterns in function addDft
```

Turning the incomplete patterns warnings in c2hs back on;

```
src/C2HS/Gen/Bind.hs:1426:5: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘addDft’:
        Patterns not matched:
            (CHSPlusParm _:_) []
            (CHSParm _ _ True _ _ _ _:_) [_]
```

I also filed a bug to fix the incomplete patterns; https://github.com/haskell/c2hs/issues/179.

--- a/c2hs.cabal
+++ b/c2hs.cabal
@@ -101,6 +101,7 @@ Extra-Source-Files:
   tests/bugs/issue-151/*.chs tests/bugs/issue-151/*.h
   tests/bugs/issue-152/*.chs tests/bugs/issue-152/*.h
   tests/bugs/issue-155/*.chs tests/bugs/issue-155/*.h
+  tests/bugs/issue-180/*.chs tests/bugs/issue-180/*.h
 
 source-repository head
   type:         git

--- a/src/C2HS/Gen/Bind.hs
+++ b/src/C2HS/Gen/Bind.hs
@@ -1438,7 +1438,11 @@ addDftMarshaller pos parms parm extTy varExTys = do
       parms'   <- addDft parms'' cts
       return (CHSParm imMarsh' hsTy True omMarsh' False p c : parms')
     addDft [] [] = return []
-    addDft ((CHSParm _ _ _ _ _ pos' _):_) [] =
+    addDft (CHSPlusParm _:_) [] =
+      marshArgMismatchErr pos "This parameter is in excess of the C arguments."
+    addDft (CHSParm _ _ _ _ _ pos' _:_) [] =
+      marshArgMismatchErr pos' "This parameter is in excess of the C arguments."
+    addDft (CHSParm _ _ True _ _ pos' _:_) [_] =
       marshArgMismatchErr pos' "This parameter is in excess of the C arguments."
     addDft [] (_:_) =
       marshArgMismatchErr pos "Parameter marshallers are missing."

--- a/tests/bugs/issue-180/Issue180.chs
+++ b/tests/bugs/issue-180/Issue180.chs
@@ -1,0 +1,13 @@
+module Main where
+
+#include "issue180.h"
+
+marshalIn = undefined
+
+{#fun pure test as test1
+  { 'marshalIn'* `Int'&} -> `()' #}
+
+main :: IO ()
+main = do
+  test
+  print "OK"

--- a/tests/bugs/issue-180/issue180.h
+++ b/tests/bugs/issue-180/issue180.h
@@ -1,0 +1,1 @@
+void test(int arg);

--- a/tests/test-bugs.hs
+++ b/tests/test-bugs.hs
@@ -92,6 +92,7 @@ tests =
     , testCase "Issue #151" issue151
     , testCase "Issue #152" issue152
     , testCase "Issue #155" issue155
+    , testCase "Issue #180" issue180
     ] ++
     -- Some tests that won't work on Windows.
     if os /= "cygwin32" && os /= "mingw32"
@@ -111,6 +112,17 @@ call_capital = c2hsShelly $ chdir "tests/bugs/call_capital" $ do
   res <- absPath "./Capital" >>= cmd
   let expected = ["upper C();", "lower c();", "upper C();"]
   liftIO $ assertBool "" (T.lines res == expected)
+
+issue180 :: Assertion
+issue180 = c2hsShelly $ chdir "tests/bugs/issue-180" $ do
+  mapM_ rm_f ["Issue180.chs.h"]
+  errExit False $ do
+    run "c2hs" [toTextIgnore "Issue180.chs"]
+  code <- lastExitCode
+  liftIO $ assertEqual "error code" 1 code
+  stderr <- lastStderr
+  let excessMsgCount = T.count "excess of the C arguments" stderr
+  liftIO $ assertBool "correct error message" (excessMsgCount == 1)
 
 issue155 :: Assertion
 issue155 = c2hsShelly $ chdir "tests/bugs/issue-155" $ do

